File: client/src/types/ai_consulting.ts
// GUI管理銘柄
export interface StockRecord {
  code: string;
  symbol: string;
  name: string | null;
  enterprise_value: number | null;
  market_cap: number | null;
  per: number | null;
  revenue: number | null;
  operating_income: number | null;
  net_income: number | null;
  dividend_yield: number | null;
  roe: number | null;
  equity_ratio: number | null;
  updated_at: string | null;
  fetched_at?: string | null;
  price_as_of?: string | null;
  financials_as_of?: string | null;
}

export interface StocksResponse {
  stocks: StockRecord[];
}

// PER
export interface PerItem {
  name: string;
  symbol: string;
  price: number | null;
  currency: string | null;
  trailing_pe: number | null;
  forward_pe: number | null;
  trailing_eps: number | null;
  computed_pe: number | null;
  pe_used: "trailing" | "forward" | "computed" | null;
  website: string | null;
  note: string | null;
}

export interface PerResponse {
  as_of: string;
  items: PerItem[];
}

// Earnings
export interface EarningsItem {
  name: string;
  symbol: string;
  earnings_date: string | null;
  price_change_5d: number | null;
  price_change_1m: number | null;
  trailing_pe: number | null;
  summary: string | null;
  website: string | null;
  note: string | null;
}

export interface EarningsWindow {
  from_date: string;
  to_date: string;
  days: number;
}

export interface EarningsResponse {
  as_of: string;
  window: EarningsWindow;
  upcoming: EarningsItem[];
  unknown: EarningsItem[];
}

// Report
export interface ReportResponse {
  as_of: string;
  per: PerResponse;
  earnings: EarningsResponse;
  text_report: string;
}

File: server/src/schemas/ai_consulting.py
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel


# Stocks (GUI管理銘柄)
class StockRecord(BaseModel):
    code: str  # 入力コード (例: "7203" or "7203.T")
    symbol: str  # 正規化済みコード: J-Quants形式 (例: "7203") / 旧yfinance形式 (例: "7203.T")
    name: Optional[str] = None  # 企業名
    enterprise_value: Optional[float] = None  # 企業価値
    market_cap: Optional[float] = None  # 時価総額
    per: Optional[float] = None  # PER
    revenue: Optional[float] = None  # 売上
    operating_income: Optional[float] = None  # 営利
    net_income: Optional[float] = None  # 純利
    dividend_yield: Optional[float] = None  # 配当利回り
    roe: Optional[float] = None  # ROE
    equity_ratio: Optional[float] = None  # 自己資本比率
    updated_at: Optional[datetime] = None


class AddStockRequest(BaseModel):
    code: str


class StocksResponse(BaseModel):
    stocks: List[StockRecord]


# PER
class PerItem(BaseModel):
    name: str
    symbol: str
    price: Optional[float] = None
    currency: Optional[str] = None
    trailing_pe: Optional[float] = None
    forward_pe: Optional[float] = None
    trailing_eps: Optional[float] = None
    computed_pe: Optional[float] = None
    pe_used: Optional[str] = None  # "trailing" | "forward" | "computed" | None
    website: Optional[str] = None
    note: Optional[str] = None


class PerResponse(BaseModel):
    as_of: datetime
    items: List[PerItem]


# Earnings
class EarningsItem(BaseModel):
    name: str
    symbol: str
    earnings_date: Optional[date] = None
    price_change_5d: Optional[float] = None
    price_change_1m: Optional[float] = None
    trailing_pe: Optional[float] = None
    summary: Optional[str] = None
    website: Optional[str] = None
    note: Optional[str] = None


class EarningsWindow(BaseModel):
    from_date: date
    to_date: date
    days: int


class EarningsResponse(BaseModel):
    as_of: datetime
    window: EarningsWindow
    upcoming: List[EarningsItem]
    unknown: List[EarningsItem]


# Report
class ReportResponse(BaseModel):
    as_of: datetime
    per: PerResponse
    earnings: EarningsResponse
    text_report: str
